src/Interpreters/ExpressionActions.cpp

```diff
@@ -47,106 +56,282 @@ namespace ErrorCodes
+{
+    if (!enable_adaptive_reorder_arguments || finished_adaptive_reorder_arguements)
+        return 0;
+    size_t n = sampled_rows.load();
+    return max_sample_rows > n ? max_sample_rows - n : 0;
+}
+
+double ExpressionShortCircuitExecuteController::calculateNodeElapsed(const ActionsDAG::Node * node)
+{
+    const auto & info = short_circuit_infos[node];
+    double res = info.profile_data.sample_rows ? info.profile_data.elapsed * 1.0 / info.profile_data.sample_rows : 0.0;
+    for (const auto * child : node->children)
     {
-        for (const auto & child : node.children)
-            result_info.nodes_info[result_info.reverse_index[child]].parents.emplace_back(&node);
+        res += calculateNodeElapsed(child);
     }
+    return res;
+}
 
-    return result_info;
+void ExpressionShortCircuitExecuteController::reorderShortCircuitFunctionsAguments()
+{
+    for (auto & [node, info] : short_circuit_infos)
+    {
+        if (node->type != ActionsDAG::ActionType::FUNCTION && node->children.size() < 2)
+            continue;
+        /// If the function is compiled, the arguments are different from the original one.
+        /// At presesnt, is_function_compiled = false, when this function has lazy executed child.
+        if (node->is_function_compiled)
+            continue;
+        IFunctionBase::ShortCircuitSettings short_circuit_settings;
+        if (!node->function_base->isShortCircuit(short_circuit_settings, node->children.size()))
+            continue;
+        if (!short_circuit_settings.support_reorder_arguments)
+            continue;
+
+        /// FIXME. A special case, select and(a,b,a) from t. if we change `and(a,b,a)` to
+        /// and(b,a,a), `a` will be executed twice. This is not a stable case, so skip it.
+        std::unordered_set<const ActionsDAG::Node *> unique_children;
+        for (const auto * child : node->children)
+        {
+            unique_children.insert(child);
+        }
+        if (unique_children.size() != node->children.size())
+            continue;
+
+        std::unordered_map<const ActionsDAG::Node *, double> node_elapsed;
+        for (const auto * child : node->children)
+        {
+            auto elpased = calculateNodeElapsed(child);
+            node_elapsed[child] = elpased;
+        }
+
+        std::vector<std::pair<UInt64, double>> args_ranks;
+        for (size_t i = 0; i < node->children.size(); ++i)
+        {
+            const auto & child_info = short_circuit_infos[node->children[i]];
+            auto child_elapsed = node_elapsed[node->children[i]];
+            double selectivity = child_info.profile_data.selected_rows * 1.0 / child_info.profile_data.sample_rows;
+            selectivity = selectivity < 0.0000001 ? 0.0000001 : selectivity; // in case it's zero.
+            if (short_circuit_settings.is_inverted_select)
+                selectivity = 1.0 / selectivity;
+            double cost = child_elapsed * selectivity;
```

Comment:
This is less optimal. Suppose we have two predicates: a and b, where a.elapsed = 0.1s, a.selectivity = 0.9, b.elapsed = 10s, b.selectivity = 0.01. a.cost = 0.45 > b.cost = 0.1. However, if a goes before b, the estimated duration would be 9.5s, otherwise it will be ~10s. 

Instead, we can define the relation < as following:

```
a.elapsed + a.selectivity * b.elapsed < b.elapsed + b.selectivity * a.elapsed

==>

a.elapsed * (1 - b.selectivity) < b.elapsed * (1 - a.selectivity)

==>

a.elapsed / (1 - a.selectivity) < b.elapsed / (1 - b.selectivity)
```

This relation is still a strict weak ordering, which can be used in std::sort.